agent/lib/kontena/websocket/client.rb

```diff
@@ -0,0 +1,739 @@
+      fail unless driver.close(reason, code) # swapped argument order
+
+      closing!
+    end
+  end
+
+#protected XXX: called by specs
+
+  # Call into driver with locked Mutex
+  #
+  # @raise [RuntimeError] not connected
+  # @yield [driver]
+  # @yieldparam driver [Websocket::Driver]
+  def with_driver
+    fail "not connected" unless @driver
+
+    @mutex.synchronize {
+      yield @driver
+    }
+  end
+
+  # Called from @driver callbacks with the @mutex held
+  #
+  # Queues block for call from #process_queue
+  def enqueue(&block)
+    fail unless block
+
+    @queue << block
+  end
+
+  # Called from read_loop without the @mutex held
+  #
+  def process_queue
+    while block = @queue.shift
+      block.call
+    end
+  end
+
+  # Connect to TCP server.
+  #
+  # @raise [Kontena::Websocket::TimeoutError] Errno::ETIMEDOUT
+  # @raise [Kontena::Websocket::ConnectError] Errno::*
+  # @return [TCPSocket]
+  def connect_tcp
+    debug "connect_tcp: timeout=#{@connect_timeout}"
+
+    Socket.tcp(self.host, self.port, connect_timeout: @connect_timeout)
+  rescue Errno::ETIMEDOUT => exc
+    raise Kontena::Websocket::TimeoutError, "Connect timeout after #{@connect_timeout}s" # XXX: actual delay
+  rescue SystemCallError => exc
+    raise Kontena::Websocket::ConnectError, exc
+  end
+
+  # @return [OpenSSL::X509::Store]
+  def ssl_cert_store
+    @ssl_cert_store ||= OpenSSL::X509::Store.new.tap do |ssl_cert_store|
+      ca_file = @ssl_params[:ca_file] || ENV['SSL_CERT_FILE']
+      ca_path = @ssl_params[:ca_path] || ENV['SSL_CERT_PATH']
+
+      if ca_file || ca_path
+        ssl_cert_store.add_file ca_file if ca_file
+        ssl_cert_store.add_path ca_path if ca_path
+      else
+        ssl_cert_store.set_default_paths
+      end
+    end
+  end
+
+  # @param ssl_cert [OpenSSL::X509::Certificate]
+  # @raise [Kontena::Websocket::SSLVerifyError]
+  def ssl_verify_cert!(ssl_cert)
+    unless ssl_cert
+      raise Kontena::Websocket::SSLVerifyError.new(OpenSSL::X509::V_OK), "No certificate"
+    end
+
+    ssl_verify_context = OpenSSL::X509::StoreContext.new(ssl_cert_store, ssl_cert)
+
+    unless ssl_verify_context.verify
+      raise Kontena::Websocket::SSLVerifyError.new(ssl_verify_context.error), "certificate verify failed: #{ssl_verify_context.error_string}"
+    end
+
+    unless OpenSSL::SSL.verify_certificate_identity(ssl_cert, self.ssl_hostname)
+      raise Kontena::Websocket::SSLVerifyError.new(OpenSSL::X509::V_OK), "Server certificate did not match hostname #{self.ssl_hostname}: #{ssl_cert.subject}"
+    end
+  end
+
+  # @param verify_result [Integer] OpenSSL::SSL::SSLSocket#verify_result
+  # @return [Kontena::Websocket::SSLVerifyError]
+  def ssl_verify_error(verify_result)
+    ssl_verify_context = OpenSSL::X509::StoreContext.new(ssl_cert_store)
+    ssl_verify_context.error = verify_result
+
+    Kontena::Websocket::SSLVerifyError.new(ssl_verify_context.error, "certificate verify failed: #{ssl_verify_context.error_string}")
```

Comment:
This is a bit of a hack with the `OpenSSL::X509::StoreContext#error=`, but it was the only way I could figure out how to call the libssl `X509_verify_cert_error_string` on the `OpenSSL::SSL::SSLSocket#verify_result` and get nice verify error messages (`self signed certificate`).